recastjs/src/recastjs.cpp

```diff
@@ -0,0 +1,577 @@
+    //
+    
+    // Once all geoemtry is rasterized, we do initial pass of filtering to
+    // remove unwanted overhangs caused by the conservative rasterization
+    // as well as filter spans where the character cannot possibly stand.
+    
+    rcFilterLowHangingWalkableObstacles(m_ctx, m_cfg.walkableClimb, *m_solid);
+    rcFilterLedgeSpans(m_ctx, m_cfg.walkableHeight, m_cfg.walkableClimb, *m_solid);
+    rcFilterWalkableLowHeightSpans(m_ctx, m_cfg.walkableHeight, *m_solid);
+    
+
+    //
+    // Step 4. Partition walkable surface to simple regions.
+    // 
+
+    // Compact the heightfield so that it is faster to handle from now on.
+    // This will result more cache coherent data as well as the neighbours
+    // between walkable cells will be calculated.
+    rcCompactHeightfield* m_chf;
+    m_chf = rcAllocCompactHeightfield();
+    if (!m_chf)
+    {
+        Log("buildNavigation: Out of memory 'chf'.");
+        return ;
+    }
+    
+    if (!rcBuildCompactHeightfield(m_ctx, m_cfg.walkableHeight, m_cfg.walkableClimb, *m_solid, *m_chf))
+    {
+        Log("buildNavigation: Could not build compact data.");
+        return ;
+    }
+    
+    if (!m_keepInterResults)
+    {
+        rcFreeHeightField(m_solid);
+        m_solid = 0;
+    }
+        
+    // Erode the walkable area by agent radius.
+    if (!rcErodeWalkableArea(m_ctx, m_cfg.walkableRadius, *m_chf))
+    {
+        Log("buildNavigation: Could not erode.");
+        return ;
+    }
+
+    // (Optional) Mark areas.
+    //class InputGeom* m_geom;
+    /*
+    const ConvexVolume* vols = m_geom->getConvexVolumes();
+    for (int i  = 0; i < m_geom->getConvexVolumeCount(); ++i)
+        rcMarkConvexPolyArea(m_ctx, vols[i].verts, vols[i].nverts, vols[i].hmin, vols[i].hmax, (unsigned char)vols[i].area, *m_chf);
+
+    if (m_monotonePartitioning)
+    {
+        // Partition the walkable surface into simple regions without holes.
+        // Monotone partitioning does not need distancefield.
+        if (!rcBuildRegionsMonotone(m_ctx, *m_chf, 0, m_cfg.minRegionArea, m_cfg.mergeRegionArea))
+        {
+            //m_ctx->log(RC_LOG_ERROR, "buildNavigation: Could not build regions.");
+            return false;
+        }
+    }
+    else
+    */
+    {
+        // Prepare for region partitioning, by calculating Distance field along the walkable surface.
+        if (!rcBuildDistanceField(m_ctx, *m_chf))
+        {
+            Log("buildNavigation: Could not build Distance field.");
+            return ;
+        }
+
+        // Partition the walkable surface into simple regions without holes.
+        if (!rcBuildRegions(m_ctx, *m_chf, 0, m_cfg.minRegionArea, m_cfg.mergeRegionArea))
+        {
+            Log("buildNavigation: Could not build regions.");
+            return ;
+        }
+    }
+    
+    //
+    // Step 5. Trace and simplify region contours.
+    //
+    
+    // Create contours.
+    rcContourSet* m_cset;
+    m_cset = rcAllocContourSet();
```

Comment:
ditto for all allocated objects in this function

Reply:
this one is deallocated actually.